onnxruntime/core/graph/contrib_ops/contrib_defs.cc

```diff
@@ -1541,6 +1589,496 @@ with the exception that numpy default keepdims to False instead of True.)DOC")
+        auto b_type = ctx.getInputType(3);
+
+        if (nullptr == a_type || nullptr == b_type ||
+            a_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType ||
+            b_type->value_case() != ONNX_NAMESPACE::TypeProto::kTensorType) {
+          fail_type_inference("inputs are expected to have tensor type.");
+        }
+
+        // validate scale and zero points
+        ValidateTypeAndShapeForScaleAndZP(ctx, 1, ONNX_NAMESPACE::TensorProto::FLOAT, true);
+        ValidateTypeAndShapeForScaleAndZP(ctx, 2, a_type->tensor_type().elem_type(), true);
+        ValidateTypeAndShapeForScaleAndZP(ctx, 4, ONNX_NAMESPACE::TensorProto::FLOAT, true);
+        ValidateTypeAndShapeForScaleAndZP(ctx, 5, b_type->tensor_type().elem_type(), true);
+        ValidateTypeAndShapeForScaleAndZP(ctx, 6, ONNX_NAMESPACE::TensorProto::FLOAT, true);
+        ValidateTypeAndShapeForScaleAndZP(ctx, 7, a_type->tensor_type().elem_type(), true);
+
+        if (hasInputShape(ctx, 0) && hasInputShape(ctx, 3)) {
+          bidirectionalBroadcastShapeInference(
+              ctx.getInputType(0)->tensor_type().shape(),
+              ctx.getInputType(3)->tensor_type().shape(),
+              *ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape());
+        }
+      });
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(QLinearAveragePool)
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .FillUsing(ONNX_NAMESPACE::PoolOpSchemaGenerator(
+          "QLinearAveragePool",
+          "average",
+          "The output of each pooling window is divided by the number of elements (exclude pad when attribute count_include_pad is zero)."
+          "Input and output scales and zero points are used to convert the output to a new quantization range."
+          "Output = Dequantize(Input) -> AveragePool on fp32 data -> Quantize(output)",
+          false))
+      .Attr(
+          "count_include_pad",
+          "Whether include pad pixels when calculating values for the edges. Default is 0, doesn't count include pad.",
+          AttributeProto::INT,
+          static_cast<int64_t>(0))
+      .Input(
+          1,
+          "x_scale",
+          "Input scale. It's a scalar, which means a per-tensor/layer quantization.",
+          "tensor(float)")
+      .Input(
+          2,
+          "x_zero_point",
+          "Input zero point. Default value is 0 if it's not specified. It's a scalar, which means a per-tensor/layer quantization.",
+          "T",
+          OpSchema::Optional)
+      .Input(
+          3,
+          "y_scale",
+          "Input B's scale. It's a scalar, which means a per-tensor/layer quantization.",
+          "tensor(float)")
+      .Input(
+          4,
+          "y_zero_point",
+          "Input B zero point. Default value is 0 if it's not specified. It's a scalar, which means a per-tensor/layer quantization.",
+          "T",
+          OpSchema::Optional);
+
+  ONNX_CONTRIB_OPERATOR_SCHEMA(FusedConvInteger)
```

Comment:
I worry about what sort of models are going to use this right away. The optimal form for int8 convolution may mean reordering the activation buffers to another format like NCHWc. This adds another step into a process that hasn't been baked yet. Can this op be deferred until later? I also wonder if activation would eventually have to be merged into this as well, but without models hard to tell yet.

Reply:
This operator fuses int32 accumulated result -> fp32 and bias addition to this fp32 result.

Any model which uses convinteger for quantization can benefit form this... For example resnet is one of them.

Having said that, if you don't think this will bring much perf gains then we can defer this for later. The most important step towards speeding resnet is adding an optimized 8 bit kernel for convolution. These fusions are only small additional steps.

Reply:
I removed this op for now.